File: platform/polyglot/confluence_api/mjs/src/services/admin-service.mjs
/**
 * @module services/admin-service
 * @description Service for Confluence Admin REST API operations.
 *
 * Provides administrative operations for user and group management:
 * creating, deleting, enabling/disabling users, setting passwords,
 * and creating/deleting groups.
 *
 * These operations require Confluence system administrator privileges.
 *
 * Confluence Data Center REST API v9.2.3.
 */

import { createLogger } from '../logger.mjs';

const log = createLogger('confluence_api', import.meta.url);

/**
 * Service class for Confluence administration operations.
 */
export class AdminService {
  /**
   * @param {import('../client/FetchClient.mjs').FetchClient} client
   *   The Confluence fetch client instance.
   */
  constructor(client) {
    /** @private */
    this._client = client;
  }

  // ---------------------------------------------------------------------------
  // User Administration
  // ---------------------------------------------------------------------------

  /**
   * Create a new user.
   *
   * @param {Object} data - User creation payload.
   * @param {string} data.username - The username for the new user.
   * @param {string} data.email - The email address for the new user.
   * @param {string} data.fullName - The full name of the new user.
   * @param {string} [data.password] - Optional initial password.
   * @returns {Promise<Object>} The created user object.
   */
  async createUser(data) {
    log.debug('createUser called', { username: data?.username, email: data?.email });
    try {
      const result = await this._client.post('admin/user', data);
      log.info('createUser succeeded', { username: data?.username });
      return result;
    } catch (error) {
      log.error('createUser failed', { username: data?.username, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Delete a user.
   *
   * @param {string} username - The username of the user to delete.
   * @returns {Promise<void>}
   */
  async deleteUser(username) {
    log.debug('deleteUser called', { username });
    try {
      const result = await this._client.delete('admin/user', {
        queryParams: { username },
      });
      log.info('deleteUser succeeded', { username });
      return result;
    } catch (error) {
      log.error('deleteUser failed', { username, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Disable a user account.
   *
   * Disabled users cannot log in but their content is preserved.
   *
   * @param {string} username - The username of the user to disable.
   * @returns {Promise<void>}
   */
  async disableUser(username) {
    log.debug('disableUser called', { username });
    try {
      const result = await this._client.post('admin/user/disable', { username });
      log.info('disableUser succeeded', { username });
      return result;
    } catch (error) {
      log.error('disableUser failed', { username, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Enable a previously disabled user account.
   *
   * @param {string} username - The username of the user to enable.
   * @returns {Promise<void>}
   */
  async enableUser(username) {
    log.debug('enableUser called', { username });
    try {
      const result = await this._client.post('admin/user/enable', { username });
      log.info('enableUser succeeded', { username });
      return result;
    } catch (error) {
      log.error('enableUser failed', { username, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Set a user's password (admin override).
   *
   * Does not require knowledge of the user's current password.
   *
   * @param {string} username - The username.
   * @param {string} password - The new password to set.
   * @returns {Promise<void>}
   */
  async setUserPassword(username, password) {
    log.debug('setUserPassword called', { username });
    try {
      const result = await this._client.put('admin/user/password', {
        username,
        password,
      });
      log.info('setUserPassword succeeded', { username });
      return result;
    } catch (error) {
      log.error('setUserPassword failed', { username, message: error.message, status: error.status });
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // Group Administration
  // ---------------------------------------------------------------------------

  /**
   * Create a new group.
   *
   * @param {string} name - The group name.
   * @returns {Promise<Object>} The created group object.
   */
  async createGroup(name) {
    log.debug('createGroup called', { name });
    try {
      const result = await this._client.post('admin/group', { name });
      log.info('createGroup succeeded', { name });
      return result;
    } catch (error) {
      log.error('createGroup failed', { name, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Delete a group.
   *
   * @param {string} groupName - The group name to delete.
   * @returns {Promise<void>}
   */
  async deleteGroup(groupName) {
    log.debug('deleteGroup called', { groupName });
    try {
      const result = await this._client.delete('admin/group', {
        queryParams: { name: groupName },
      });
      log.info('deleteGroup succeeded', { groupName });
      return result;
    } catch (error) {
      log.error('deleteGroup failed', { groupName, message: error.message, status: error.status });
      throw error;
    }
  }
}

File: platform/polyglot/confluence_api/py/src/confluence_api/services/admin_service.py
"""
Admin service for Confluence Data Center REST API v9.2.3.

Provides administrative operations for user and group management
including creation, deletion, enable/disable, and password resets.
"""

from __future__ import annotations

from typing import Any

from confluence_api.logger import create_logger

log = create_logger("confluence_api", __file__)


class AdminService:
    """Service for Confluence administrative operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    # ── Admin User ───────────────────────────────────────────────────────

    def create_user(self, data: dict) -> dict:
        """
        Create a new user.

        POST /rest/api/admin/user

        Args:
            data: User creation payload containing username, email, fullName, etc.

        Returns:
            Created user data dict.
        """
        log.debug("create_user called", {"username": data.get("username")})
        result = self._client.post("admin/user", json_data=data)
        log.info("create_user succeeded", {"username": data.get("username")})
        return result

    def delete_user(self, username: str) -> dict:
        """Delete a user by username."""
        log.debug("delete_user called", {"username": username})
        result = self._client.delete("admin/user", params={"username": username})
        log.info("delete_user succeeded", {"username": username})
        return result

    def disable_user(self, username: str) -> dict:
        """Disable a user account."""
        log.debug("disable_user called", {"username": username})
        result = self._client.post(
            "admin/user/disable", json_data={"username": username}
        )
        log.info("disable_user succeeded", {"username": username})
        return result

    def enable_user(self, username: str) -> dict:
        """Enable a previously disabled user account."""
        log.debug("enable_user called", {"username": username})
        result = self._client.post(
            "admin/user/enable", json_data={"username": username}
        )
        log.info("enable_user succeeded", {"username": username})
        return result

    def set_user_password(self, username: str, password: str) -> dict:
        """Set a user's password (admin operation)."""
        log.debug("set_user_password called", {"username": username})
        result = self._client.post(
            "admin/user/password",
            json_data={"username": username, "password": password},
        )
        log.info("set_user_password succeeded", {"username": username})
        return result

    # ── Admin Group ──────────────────────────────────────────────────────

    def create_group(self, name: str) -> dict:
        """Create a new group."""
        log.debug("create_group called", {"name": name})
        result = self._client.post("admin/group", json_data={"name": name})
        log.info("create_group succeeded", {"name": name})
        return result

    def delete_group(self, group_name: str) -> dict:
        """Delete a group by name."""
        log.debug("delete_group called", {"group_name": group_name})
        result = self._client.delete(
            "admin/group", params={"name": group_name}
        )
        log.info("delete_group succeeded", {"group_name": group_name})
        return result


class AsyncAdminService:
    """Async service for Confluence administrative operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    # ── Admin User ───────────────────────────────────────────────────────

    async def create_user(self, data: dict) -> dict:
        """Create a new user."""
        log.debug("create_user called", {"username": data.get("username")})
        result = await self._client.post("admin/user", json_data=data)
        log.info("create_user succeeded", {"username": data.get("username")})
        return result

    async def delete_user(self, username: str) -> dict:
        """Delete a user by username."""
        log.debug("delete_user called", {"username": username})
        result = await self._client.delete("admin/user", params={"username": username})
        log.info("delete_user succeeded", {"username": username})
        return result

    async def disable_user(self, username: str) -> dict:
        """Disable a user account."""
        log.debug("disable_user called", {"username": username})
        result = await self._client.post(
            "admin/user/disable", json_data={"username": username}
        )
        log.info("disable_user succeeded", {"username": username})
        return result

    async def enable_user(self, username: str) -> dict:
        """Enable a previously disabled user account."""
        log.debug("enable_user called", {"username": username})
        result = await self._client.post(
            "admin/user/enable", json_data={"username": username}
        )
        log.info("enable_user succeeded", {"username": username})
        return result

    async def set_user_password(self, username: str, password: str) -> dict:
        """Set a user's password (admin operation)."""
        log.debug("set_user_password called", {"username": username})
        result = await self._client.post(
            "admin/user/password",
            json_data={"username": username, "password": password},
        )
        log.info("set_user_password succeeded", {"username": username})
        return result

    # ── Admin Group ──────────────────────────────────────────────────────

    async def create_group(self, name: str) -> dict:
        """Create a new group."""
        log.debug("create_group called", {"name": name})
        result = await self._client.post("admin/group", json_data={"name": name})
        log.info("create_group succeeded", {"name": name})
        return result

    async def delete_group(self, group_name: str) -> dict:
        """Delete a group by name."""
        log.debug("delete_group called", {"group_name": group_name})
        result = await self._client.delete(
            "admin/group", params={"name": group_name}
        )
        log.info("delete_group succeeded", {"group_name": group_name})
        return result
